haskell/import.bzl

```diff
@@ -4,32 +4,71 @@ load(":private/context.bzl", "haskell_context")
 load(":private/actions/package.bzl", "package")
 load(
     ":private/providers.bzl",
+    "HaddockInfo",
     "HaskellBuildInfo",
     "HaskellLibraryInfo",
 )
 load(":private/set.bzl", "set")
 load(":private/path_utils.bzl", "ln")
 load("@bazel_skylib//:lib.bzl", "paths")
 
+def link_forest(ctx, srcs, basePath = ".", **kwargs):
+    """Write a symlink to each file in `srcs` into a destination directory
+    defined using the same arguments as `ctx.actions.declare_directory`"""
+    local_files = []
+    for src in srcs:
+        dest = ctx.actions.declare_file(
+            paths.join(basePath, src.basename),
+            **kwargs
+            )
+        local_files.append(dest)
+        ln(ctx, src, dest)
+    return local_files
+
+def copy_all(ctx, srcs, dest):
+    """Copy all the files in `srcs` into `dest`"""
+    if list(srcs) == []:
+      ctx.actions.run_shell(
+        command = "mkdir -p {dest}".format(dest = dest.path),
+        outputs = [dest],
+      )
+    else:
+        args = ctx.actions.args()
+        args.add_all(srcs)
+        ctx.actions.run_shell(
+            inputs = depset(srcs),
+            outputs = [dest],
+            mnemonic = "Copy",
+            command = "mkdir -p {dest} && cp -L -R \"$@\" {dest}".format(dest = dest.path),
+            arguments = [args],
+        )
+
 def _haskell_import_impl(ctx):
     hs = haskell_context(ctx)
 
     package_cache = ctx.actions.declare_file(
-        paths.join(ctx.attr.package_id, "package.conf.d", "package.cache"),
+        paths.join("package.conf.d", "package.cache")
+    )
+
+    local_package_confs = link_forest(
+        ctx =ctx,
+        srcs = ctx.attr.package_conf.files,
+        sibling = package_cache,
     )
 
-    local_package_confs = []
-    for package_conf_file in ctx.attr.package_conf.files:
-        local_package_conf = ctx.actions.declare_file(package_conf_file.basename, sibling = package_cache)
-        local_package_confs.append(local_package_conf)
-        ln(ctx, package_conf_file, local_package_conf)
 
-    package_path = package_cache.dirname + ":"
-    recache_args = ctx.actions.args()
-    recache_args.add(["-D"] + list(ctx.attr.package_conf.files) + [
-        "--target-directory",
-        local_package_conf,
-    ])
+    local_haddock_html = ctx.actions.declare_directory("haddock-html")
+    copy_all(
+        ctx = ctx,
+        srcs = ctx.attr.haddock_html.files,
+        dest = local_haddock_html,
+    )
```

Comment:
I don't understand why (probably related to the internal cooking of the `haddock` aspect), but just symlinking these caused me some permission errors when trying to build the haddocks, hence the copying

Reply:
Hm, interesting. Can you paste the error messages? I didn’t manage to change it to a symlink semantics (naïvely switching to `link_forest` doesn’t work).

Maybe this also warrants a code comment.

Reply:
By changing the `copy_all` rule as

```diff
diff --git a/haskell/private/path_utils.bzl b/haskell/private/path_utils.bzl
index 7126118..8fd2e2f 100644
--- a/haskell/private/path_utils.bzl
+++ b/haskell/private/path_utils.bzl
@@ -237,11 +237,12 @@ def copy_all(ctx, srcs, dest):
         )
     else:
         args = ctx.actions.args()
-        args.add_all(srcs)
+        real_srcs = [ _truly_relativize(src.path, dest.path) for src in srcs]
+        args.add_all(real_srcs)
         ctx.actions.run_shell(
             inputs = depset(srcs),
             outputs = [dest],
             mnemonic = "Copy",
-            command = "mkdir -p {dest} && cp -L -R \"$@\" {dest}".format(dest = dest.path),
+            command = "mkdir -p {dest} && ln -s \"$@\" {dest}".format(dest = dest.path),
             arguments = [args],
         )
```

(so which effectively does a symlinking of everything), I get

```
ERROR: /home/regnat-dev/.cache/bazel/_bazel_regnat-dev/8da26023ce322171e002355d3467f67c/external/hackage-template-haskell/BUILD:6:1: Error while validating output Tree
Artifact File:[[<execution_root>]bazel-out/k8-fastbuild/bin]external/hackage-template-haskell/haddock-html : /home/regnat-dev/.cache/bazel/_bazel_regnat-dev/8da26023ce
322171e002355d3467f67c/execroot/io_tweag_rules_haskell/bazel-out/k8-fastbuild/bin/external/hackage-template-haskell/haddock-html/template-haskell-2.12.0.0 (Read-only f
ile system)
```

I must admit I didn't try to dig deep into this since copying is cheap enough not to be a real problem